### PR TITLE
Add Cost Effective Gradient Boosting

### LINKED ISSUE #1119: Adding Cost Effective Gradient Boosting to LightGBM
S Peter proposes a feature which can be added to LightGBM as a feature. The work has already been done in a fork here- https://github.com/svenpeter42/LightGBM-CEGB

Paper related to the topic- https://papers.nips.cc/paper/6753-cost-efficient-gradient-boosting.pdf

### PR DESCRIPTION
**Pull request by remcob-gr**
Fixes #1119 .
The implementation is in the form of a tweak to the serial tree learner and so should work with every derived tree learner, though I've only tested in serial.

### CODE REVIEW

**Comment by guolinke** [comment]
is this ready to merge ?

**Comment by remcob-gr** [comment]
@StrikerRUS: It's ready from my perspective.
@guolinke : Could you review?

File: src/treelearner/serial_tree_learner.cpp [hunk 1]
```diff
@@ -703,6 +753,26 @@ int32_t SerialTreeLearner::ForceSplits(Tree* tree, Json& forced_split_json, int*
 void SerialTreeLearner::Split(Tree* tree, int best_leaf, int* left_leaf, int* right_leaf) {
   const SplitInfo& best_split_info = best_split_per_leaf_[best_leaf];
   const int inner_feature_index = train_data_->InnerFeatureIndex(best_split_info.feature);
+  if(!config_->cegb_penalty_feature_coupled.empty() && !feature_used[inner_feature_index]){
+    feature_used[inner_feature_index] = true;
+    for(int i = 0; i < tree->num_leaves(); i++){
```
Comment (StrikerRUS) [inline src/treelearner/serial_tree_learner.cpp]: `++i` for the consistency with other codebase.

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -377,6 +377,28 @@ struct Config {
   // desc = used only in ``refit`` task in CLI version or as argument in ``refit`` function in language-specific package
   double refit_decay_rate = 0.9;
 
+  // check = >=0.0
+  // default = 1
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: This line can be removed.

File: include/LightGBM/config.h [hunk 3]
```diff
@@ -377,6 +377,28 @@ struct Config {
...
+  // desc = cost-effective gradient boosting multiplier for all penalties
+  double cegb_tradeoff = 1.0;
+
+  // check = >=0.0
+  // default = 0
```
Comment (StrikerRUS) [inline include/LightGBM/config.h]: This line can be removed.

**Comment by StrikerRUS** [comment]
I think we should cite CEGB somehow...
Maybe in params description, like `` `cost-effective gradient-boosting` <https://papers.nips.cc/paper/6753-cost-efficient-gradient-boosting.pdf>__ penalty for ...`` ?

File: src/treelearner/serial_tree_learner.cpp [hunk 4]
```diff
@@ -496,6 +530,14 @@ void SerialTreeLearner::FindBestSplitsFromHistograms(const std::vector<int8_t>&
       smaller_leaf_splits_->max_constraint(),
       &smaller_split);
     smaller_split.feature = real_fidx;
+    smaller_split.gain -= config_->cegb_tradeoff * config_->cegb_penalty_split * smaller_leaf_splits_->num_data_in_leaf();
```
Comment (guolinke) [inline src/treelearner/serial_tree_learner.cpp]: `config_->cegb_tradeoff * config_->cegb_penalty_split` is zero by default, right?

**Comment by guolinke** [comment]
I think it will be better to add a section into advaced-topics: https://github.com/Microsoft/LightGBM/blob/master/docs/Advanced-Topics.rst, about how to use cfgb.

File: src/treelearner/serial_tree_learner.cpp (hunk 4 above)
Comment (remcob-gr) [reply src/treelearner/serial_tree_learner.cpp]: Yes. By default, this line doesn't change the gain, and so doesn't change the behaviour.
It only makes a difference if the user specifies a cegb_penalty_split.

**Comment by remcob-gr** [comment]
I've added a section to the docs on using CEGB, including a link to the paper.
@StrikerRUS , @guolinke : Could you take a look?

**Comment by guolinke** [comment]
Thanks @remcob-gr ,it looks good to me.

**Comment by remcob-gr** [comment]
@StrikerRUS : I've added some tests. Could you take a look and see if there are other tests you'd like?

**Comment by StrikerRUS** [comment]
@remcob-gr Perfect! Many thanks!

@guolinke Can we merge?

**Comment by guolinke** [comment]
@StrikerRUS sure

**Pull request merged**